src/lib/png/png.c

```diff
@@ -0,0 +1,1128 @@
+
+				if(i == 0) {
+					err = -1; /* can't repeat previous if i is 0 */
+					break;
+				}
+
+				if((*bp + 2) > in_bitlen) {
+					err = -1; /* error, bit pointer jumps past memory */
+					break;
+				}
+				replength += readBitsFromStream(bp, in, 2);
+
+				if(i < HLIT + 1) {
+					value = bitlen_ll[i - 1];
+				} else {
+					value = bitlen_d[i - HLIT - 1];
+				}
+				/* repeat this value in the next lengths */
+				for(n = 0; n < replength; n++) {
+					if(i >= HLIT + HDIST) {
+						err = -1; /* error: i is larger than the amount of codes */
+						break;
+					}
+					if(i < HLIT) {
+						bitlen_ll[i] = value;
+					} else {
+						bitlen_d[i - HLIT] = value;
+					}
+					i++;
+				}
+			} else if(code == 17) /* repeat "0" 3-10 times */ {
+				replength = 3; /* read in the bits that indicate repeat length */
+
+				if((*bp + 3) > in_bitlen) {
+					err = -1; /* error, bit pointer jumps past memory */
+					break;
+				}
+
+				replength += readBitsFromStream(bp, in, 3);
+
+				/* repeat this value in the next lengths */
+				for(n = 0; n < replength; n++) {
+					if(i >= HLIT + HDIST) {
+						return -1; /* error: i is larger than the amount of codes */
+					}
+
+					if(i < HLIT) {
+						bitlen_ll[i] = 0;
+					} else {
+						bitlen_d[i - HLIT] = 0;
+					}
+					i++;
+				}
+			} else if(code == 18) /* repeat "0" 11-138 times */ {
+				replength = 11; /* read in the bits that indicate repeat length */
+				if((*bp + 7) > in_bitlen) {
+					err = -1; /* error, bit pointer jumps past memory */
+					break;
+				}
+				replength += readBitsFromStream(bp, in, 7);
+
+				/* repeat this value in the next lengths */
+				for(n = 0; n < replength; n++) {
+					if(i >= HLIT + HDIST) {
+						err = -1; /* error: i is larger than the amount of codes */
+						break;
+					}
+					if(i < HLIT) {
+						bitlen_ll[i] = 0;
+					} else {
+						bitlen_d[i - HLIT] = 0;
+					}
+					i++;
+				}
+			} else {
+				if(code == -1) {
+					err = -1; /* error in decodesymbol fun */
+					break;
+				}
+			}
+		}
+		if(err) break;
+
+		if(bitlen_ll[256] == 0) {
+			err = -1; /* the length of the end code 256 must be larger than 0 */
+			break;
+		}
+
+		/* generate the code trees */
+		err = make_tree_first_step(ll_tree, bitlen_ll, NUMBER_OF_CODE_SYMBOLS, 15);
+		if(err) break;
+		err = make_tree_first_step(dist_tree, bitlen_d, NUMBER_OF_DISTANCE_SYMBOLS, 15);
+
+		break; /* end of while(!err) */
+	}
+	HuffmanTree_cleanup(&cl_tree);
+	return err;
+}
+
+/* get the dist_tree with fixed huffman */
+int build_FixeddTree(HuffmanTree *tree) {
+	int i, err = 0;
+	int bitlen[NUMBER_OF_DISTANCE_SYMBOLS];
+
+	for(i = 0; i != NUMBER_OF_DISTANCE_SYMBOLS; i++) {
+		bitlen[i] = 5;
+	}
+
+	err = make_tree_first_step(tree, bitlen, NUMBER_OF_DISTANCE_SYMBOLS, 15);
+	return err;
+}
+
+/* get the ll_tree with fixed huffman */
+int build_FixedllTree(HuffmanTree *tree) {
+	int i, err = 0;
+	int bitlen[NUMBER_OF_CODE_SYMBOLS];
+
+	for(i = 0; i <= 143; i++) {
```

Comment:
What are these loop boundaries: 143, 255, 279, 287? Where do they come from?

Reply:
they are from table for fixed huffman codes and they represent the boundaries of the unpacked values

Reply:
if you are interested to delve into this, here is the easiest option(https://habr.com/en/post/274825/) for a deeper understanding need to see rfc1951

Reply:
I mean: reader of the code should understand what's going on, so you should either replace such constants with meaningful macros or add comments.

Reply:
oh, okey, for some reason I thought you was just wondering

Reply:
add comments in this places